apisix/plugins/authz-keycloak.lua

```diff
@@ -64,10 +66,164 @@ local _M = {
     schema = schema,
 }
 
+
 function _M.check_schema(conf)
+    if not conf.discovery and not conf.token_endpoint then
+        return false, 'Neither discovery nor token endpoint given.'
+    end
     return core.schema.check(schema, conf)
 end
 
+
+-- Retrieve value from server-wide cache, if available.
+local function authz_keycloak_cache_get(type, key)
+  local dict = ngx.shared[type]
+  local value
+  if dict then
+    value = dict:get(key)
+    if value then log.debug("cache hit: type=", type, " key=", key) end
+  end
+  return value
+end
+
+
+-- Set value in server-wide cache, if available.
+local function authz_keycloak_cache_set(type, key, value, exp)
+  local dict = ngx.shared[type]
+  if dict and (exp > 0) then
+    local success, err, forcible = dict:set(key, value, exp)
```

Comment:
Better to `log.error` if set dict failed.

Reply:
Fixed.